### PR TITLE
EOS-18618 - Fix memory leaks from evhttp_uridecode and replace deprecated evhttp_decode_uri

### PR DESCRIPTION
**Pull request by shahidSTX**
EOS-18618 - Fix memory leaks from evhttp_uridecode and replace deprecated evhttp_decode_uri

Files modified :
   server/s3_copy_object_action.cc
   server/s3_post_multipartobject_action.cc
   server/s3_put_chunk_upload_object_action.cc
   server/s3_put_object_action.cc

Testing :
On CLI :
`put-object pbucket1/object - result(PASS)`
`copy-object pbucket1/object - result(PASS)`
End result : put-object and copy-object via CLI is successful for regular objects.

Signed-off-by: Mohammed Shahid <mohammed.shahid@seagate.com>

### CODE REVIEW

File: server/s3_post_multipartobject_action.cc [hunk 1]
```diff
@@ -171,7 +171,7 @@ void S3PostMultipartObjectAction::parse_x_amz_tagging_header(
     for (struct evkeyval* header = key_value.tqh_first; header;
          header = header->next.tqe_next) {
 
-      decoded_key = evhttp_decode_uri(header->key);
+      decoded_key = evhttp_uridecode(header->key, 0, NULL);
```
Comment (evgeniy-brazhnikov) [inline server/s3_post_multipartobject_action.cc:174]: "The returned string must be freed by the caller."
Did you forget free() ?

Comment (shahidSTX) [reply server/s3_post_multipartobject_action.cc:174]: No, decoded_key is allocated memory from heap and is used by :
`new_object_tags_map[decoded_key] = header->value;`
This new_object_tags_map is used by validate_object_xml_tags and lives throughout the duration of the program, so we cannot free it. I tried freeing it which was causing failures.

Comment (evgeniy-brazhnikov) [reply server/s3_post_multipartobject_action.cc:174]: It's strange. A compiler must create temporary std::string object for the expression new_object_tags_map[decoded_key].
And decoded_key variable must be useless just after the expression.
Can you please retest

Comment (shahidSTX) [reply server/s3_post_multipartobject_action.cc:174]: new_object_tags_map is a std::map, not a temporary string, and decoded_key is one of its mapped entry.
I did testing couple of times and saw failures every time I tried to free decoded_key.

Comment (shahidSTX) [reply server/s3_post_multipartobject_action.cc:174]: Done, freeing up after the expression works.

File: server/s3_copy_object_action.cc [hunk 2]
```diff
@@ -91,7 +91,10 @@ void S3CopyObjectAction::get_source_bucket_and_object() {
       source_object_name = source.substr(separator_pos + 1);
     }
   }
-  source_object_name = evhttp_decode_uri(source_object_name.c_str());
+  char* decode_uri = evhttp_uridecode(source_object_name.c_str(), 0, NULL);
+  source_object_name = decode_uri;
+  free(decode_uri);
+  decode_uri = NULL;
```
Comment (evgeniy-brazhnikov) [inline server/s3_copy_object_action.cc:97]: For future: it's redundant in such cases. The variable isn't used later. Moreover it get out of scope on the next line.

**Review by evgeniy-brazhnikov — APPROVED** [review]

**Review by amitwac2608 — APPROVED** [review]

File: server/s3_put_chunk_upload_object_action.cc [hunk 3]
```diff
@@ -208,11 +208,13 @@ void S3PutChunkUploadObjectAction::parse_x_amz_tagging_header(
     for (struct evkeyval* header = key_value.tqh_first; header;
          header = header->next.tqe_next) {
 
-      decoded_key = evhttp_decode_uri(header->key);
+      decoded_key = evhttp_uridecode(header->key, 0, NULL);
```
Comment (bkirunge7) [inline server/s3_put_chunk_upload_object_action.cc:211]: does this set decoded_key to null if evhttp_uriecode fails

Comment (shahidSTX) [reply server/s3_put_chunk_upload_object_action.cc:211]: Yes.

File: server/s3_copy_object_action.cc (hunk 2 above)
Comment (knrajnambiar76) [inline server/s3_copy_object_action.cc:97]: Not needed to make it NULL as decode_uri is local

**Review by knrajnambiar76 — APPROVED** [review]

**Pull request merged**